### PR TITLE
Add Forbidden Classes sniff

### PR DESCRIPTION
**Pull request by arxeiss**
Closing #962 

Added sniff to disable used classes, interfaces, base classes or traits.

### CODE REVIEW

File: SlevomatCodingStandard/Helpers/NamespaceHelper.php [hunk 1]
```diff
@@ -45,6 +45,15 @@ public static function isFullyQualifiedName(string $typeName): bool
 		return StringHelper::startsWith($typeName, self::NAMESPACE_SEPARATOR);
 	}
 
+	public static function getFullyQualifiedTypeName(string $typeName): string
```
Comment (kukulich) [inline SlevomatCodingStandard/Helpers/NamespaceHelper.php:48]: Please move this method and test to separate commit.

File: tests/Sniffs/PHP/data/forbiddenClassesErrors.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		};
+
+		function (Dododo &$dododo) : array {
+			return $dododo;
+		};
+	}
+
+}
+
+// interface FooInterface extends \ExtendedInterface, \SecondExtendedInterface, \ThirdExtendedInterface
```
Comment (kukulich) [inline tests/Sniffs/PHP/data/forbiddenClassesErrors.php:57]: WTF?

File: tests/Sniffs/PHP/ForbiddenClassesSniffTest.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		self::assertSniffError($report, 29, ForbiddenClassesSniff::CODE_FORBIDDEN_INTERFACE);
+		self::assertSniffError($report, 31, ForbiddenClassesSniff::CODE_FORBIDDEN_TRAIT);
+		self::assertSniffError($report, 38, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 40, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 43, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 47, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 48, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertCount(2, $allErrors[48]);
+		self::assertSniffError($report, 49, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertCount(2, $allErrors[49]);
```
Comment (kukulich) [inline tests/Sniffs/PHP/ForbiddenClassesSniffTest.php:57]: You should check specific error messages (forth parameter of `assertSniffError()`) - this is useful only on line 48 where both errors are same. The same review is valid for line 9.

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+			if ($alternative === null) {
+				$phpcsFile->addError(
+					sprintf('Usage of %s %s is forbidden.', $reference->fullyQualifiedName, $nameType),
+					$reference->startPointer,
+					$code
+				);
+			} elseif (!$isFixable) {
+				$phpcsFile->addError(
+					sprintf(
+						'Usage of %s %s if forbidden, use %s instead.',
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php:179]: Typo.

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+						$nameType,
+						$alternative
+					),
+					$reference->startPointer,
+					$code
+				);
+			} else {
+				$fix = $phpcsFile->addFixableError(
+					sprintf(
+						'Usage of %s %s if forbidden, use %s instead.',
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php:190]: Typo.

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+					$code
+				);
+				if (!$fix) {
+					continue;
+				}
+
+				$phpcsFile->fixer->beginChangeset();
+				$phpcsFile->fixer->replaceToken($reference->startPointer, $alternative);
+				for ($i = $reference->startPointer + 1; $i <= $reference->endPointer; ++$i) {
+					$phpcsFile->fixer->replaceToken((int) $i, '');
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php:205]: `(int)` looks strange and should not be needed.

File: SlevomatCodingStandard/Helpers/NamespaceHelper.php (hunk 1 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Helpers/NamespaceHelper.php:48]: Not sure I understand. I thought I should squash everything into 1 commit. You ask to separate it. Does it mean 2 commits in this PR, or 2 separate PRs?

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php (hunk 6 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php:205]: I also thought, but then I got error from PHP Stan, that replace token expects `int`, but `int|float` is receiving. Will see if it disappear after rebase.

File: tests/Sniffs/PHP/data/forbiddenClassesErrors.php (hunk 2 above)
Comment (arxeiss) [reply tests/Sniffs/PHP/data/forbiddenClassesErrors.php:57]: Oh crap, this file is not used anymore, forgot to remove.

File: SlevomatCodingStandard/Helpers/NamespaceHelper.php (hunk 1 above)
Comment (kukulich) [reply SlevomatCodingStandard/Helpers/NamespaceHelper.php:48]: This change can be in separate commit in this PR, so in this PR will be 2 commit. Changes in `NamespaceHelper` and the new sniff.

File: README.md [hunk 7]
```diff
@@ -298,6 +298,30 @@ Looks for `use` alias that is same as unqualified name.
 
 Disallows references.
 
+#### SlevomatCodingStandard.PHP.ForbiddenClasses 🔧
+
+Sniff checks forbidden classes, interfaces, parent classes and traits. And provide the following settings:
+
+- `forbiddenClasses` - Creating instances with `new` keyword or accessing with `::` operator
```
[the remark this answers is not in the document]
Comment (arxeiss) [reply README.md]: Fixed

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+	public $forbiddenExtends = [];
+
+	/** @var array<string, (string|null)> */
+	public $forbiddenInterfaces = [];
+
+	/** @var array<string, (string|null)> */
+	public $forbiddenTraits = [];
+
+	/** @var array<string> */
+	public static $keywordReferences = ['self', 'parent', 'static'];
```
[the remark this answers is not in the document]
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: Fixed

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+		}
+
+		return $references;
+	}
+
+	/**
+	 * @param array<string, (string|null)> $param
+	 * @return array<string, (string|null)>
+	 */
+	private static function normalizeInputParam(array $param): array
```
[the remark this answers is not in the document]
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: Changed into `option`

File: SlevomatCodingStandard/Helpers/NamespaceHelper.php (hunk 1 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Helpers/NamespaceHelper.php:45]: Split into 2 commits

File: tests/Sniffs/PHP/data/forbiddenClassesErrors.php (hunk 2 above)
Comment (arxeiss) [reply tests/Sniffs/PHP/data/forbiddenClassesErrors.php]: File was removed

File: tests/Sniffs/PHP/ForbiddenClassesSniffTest.php (hunk 3 above)
Comment (arxeiss) [reply tests/Sniffs/PHP/ForbiddenClassesSniffTest.php]: Changed for message check expect line 48

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php (hunk 4 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: Fixed

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php (hunk 5 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: Fixed

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php (hunk 6 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: I tried to remove `(int)` but this errors is shown: https://travis-ci.org/github/slevomat/coding-standard/jobs/672883870

Comment (kukulich) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: That's probably because of `min()` call and should be fixed there (via `@var` annotation)

Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: No it is not. I tried to put there `/** @var int $nextSeparator */` and errors remains.

Also if I change the line into this, which does not make sense, then PHP Stan pass. Only with the `$i` variable doesn't pass.

```php
// this works, when removed + 1 works also
$phpcsFile->fixer->replaceToken($reference->startPointer + 1, ''); 
```
I also tried to add `/** @var $i int */` into the for loop. This works, but PHP CS fails. with `No assignment to $i variable before or after the documentation comment.`

Comment (kukulich) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: Ok, another try:

Try to change `array<stdClass>` to array shape, see https://github.com/phpstan/phpdoc-parser/pull/30

So modify `stdClass` to array.

Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: Changing stdClass to an array helped. Please re-review

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+			if ($alternative === null) {
+				$phpcsFile->addError(
+					sprintf('Usage of %s %s is forbidden.', $reference['fullyQualifiedName'], $nameType),
+					$reference['startPointer'],
+					$code
+				);
+			} elseif (!$isFixable) {
+				$phpcsFile->addError(
+					sprintf(
+						'Usage of %s %s of forbidden, use %s instead.',
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php:178]: New typo :)

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+						$nameType,
+						$alternative
+					),
+					$reference['startPointer'],
+					$code
+				);
+			} else {
+				$fix = $phpcsFile->addFixableError(
+					sprintf(
+						'Usage of %s %s of forbidden, use %s instead.',
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php:189]: New typo :)

File: tests/Sniffs/PHP/ForbiddenClassesSniffTest.php [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		self::assertSniffError($report, 40, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 43, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 47, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertSniffError($report, 48, ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS);
+		self::assertCount(2, $allErrors[48]);
+		self::assertSniffError(
+			$report,
+			49,
+			ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS,
+			'Usage of \DB class of forbidden, use \TypeORM\DB instead'
```
Comment (kukulich) [inline tests/Sniffs/PHP/ForbiddenClassesSniffTest.php:69]: The typo again :)

File: tests/Sniffs/PHP/ForbiddenClassesSniffTest.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+		);
+		self::assertSniffError(
+			$report,
+			49,
+			ForbiddenClassesSniff::CODE_FORBIDDEN_CLASS,
+			'Usage of \UserModel class of forbidden, use \MyApp\Models\User instead'
```
Comment (kukulich) [inline tests/Sniffs/PHP/ForbiddenClassesSniffTest.php:75]: And again :)

File: SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php (hunk 10 above)
Comment (arxeiss) [reply SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php]: :man_facepalming: sorry

**Comment by kukulich** [comment]
Thanks.
